### ISSUE TITLE
Let users specify use_frameworks! in their Podfile

### ISSUE THREAD

**Issue by jcanizales** [grpc/grpc#4121]
As reported by @bdotdub in https://github.com/grpc/grpc/pull/4108#issuecomment-156303012, pods written in Swift require `use_frameworks!` in the user's Podfile. gRPC's podspec isn't currently compatible with that, IIRC because of OpenSSL.

- [x] Make OpenSSL's podspec compatible with frameworks (we have to fix it anyway to solve #4081).
- [ ] Use a pod written in Swift in the SwiftSample, to prevent a regression.

**Comment by Gurpartap** [comment]
#4235 is letting me specify `use_frameworks!`.

```ruby
source "https://github.com/CocoaPods/Specs.git"
platform :ios, "8.0"

use_frameworks!
inhibit_all_warnings!

pod "Protobuf", "~> 3.0.0-alpha-4.1"
pod "BoringSSL"
pod "gRPC", git: "https://github.com/grpc/grpc" // 0.12 wasn't out at the time.
pod "GRPCAPISDK", path: "somewhere" // E_TOO_MANY_CAPS.
```

**Comment by Gurpartap** [comment]
However, the grpc implementation pod (`GRPCAPI` in my case) is an empty group. It's a direct port of the route_guide.

![screen shot 2015-12-09 at 6 26 43 am](https://cloud.githubusercontent.com/assets/39792/11673323/d716b62a-9e3d-11e5-97c6-2e0387d05f04.png)

```sh
$ bundle exec pod --version
0.39.0
```

```ruby
$ cat GRPCAPI.podspec
Pod::Spec.new do |s|
  s.name     = "GRPCAPI"
  s.version  = "0.0.1"
  s.license  = "New BSD"

  s.ios.deployment_target = "7.1"
  s.osx.deployment_target = "10.9"

  # Base directory where the .proto files are.
  src = "/path/to/protos"

  # Directory where the generated files will be placed.
  dir = "/path/to/Pods/" + s.name

  # Run protoc with the Objective-C and gRPC plugins to generate protocol messages and gRPC clients.
  s.prepare_command = <<-CMD
    mkdir -p #{dir}
    protoc -I #{src} --objc_out=#{dir} --objcgrpc_out=#{dir} #{src}/*.proto
  CMD

  s.subspec "Messages" do |ms|
    ms.source_files = "#{dir}/*.pbobjc.{h,m}", "#{dir}/**/*.pbobjc.{h,m}"
    ms.header_mappings_dir = dir
    ms.requires_arc = false
    ms.dependency "Protobuf", "~> 3.0.0-alpha-4"
  end

  s.subspec "Services" do |ss|
    ss.source_files = "#{dir}/*.pbrpc.{h,m}", "#{dir}/**/*.pbrpc.{h,m}"
    ss.header_mappings_dir = dir
    ss.requires_arc = true
    ss.dependency "gRPC", "~> 0.12"
    ss.dependency "#{s.name}/Messages"
  end
end
```

**Comment by Gurpartap** [comment]
Files not being included is not related to this issue since I have this even without `use_frameworks!` and OpenSSL pod instead of BoringSSL.

I'll create another issue.

**Comment by matti777** [comment]
At least for me -- using CocoaPods 1.0.0.beta.6 -- the headers still are not created and this does not compile as is.

**Comment by fabware** [comment]
After half of year, it JUST can't work, disappointing.

**Comment by jcanizales** [comment]
This is blocked by https://github.com/CocoaPods/CocoaPods/issues/4605, which hasn't seen much action. Until the 1.0.0 release, Cocoapods itself was a moving ground, so this has been quite the challenge for us to debug too.

@timburks started working on the idea to use [Bazel](www.bazel.io) to precompile a dynamic framework, and just have Cocoapods vendor it. This brought new challenges, and is not ultimately what we'll do. But it was essential to let me move this forward: Bazel generates an XCode project that's able to correctly compile a framework for gRPC, at least w.r.t. the header locations. This allowed me to study the XCode projects output by both Bazel and Cocoapods, to find how Cocoapods _should_ do it.

**Comment by jcanizales** [comment]
(This study is valid as of version 0.2.2b of Bazel).

From `BUILD` rules like these:

```python
ios_framework(
  name = "grpc",
  hdrs = [],
  binary = ":MyFramework_binary",
)

ios_framework_binary(
  name = "MyFramework_binary",
  hdrs = [],
  includes = ["include"],
  deps = [
    ":MyFramework_library",
  ],
)

objc_library(
    name = "MyFramework_library",
    srcs = glob([
      "nested/dir/mynestedsource.c",
      "nested/dir/mynestedheader.h",
    ]),
    hdrs = glob([
      "include/grpc/myheader.h",
    ]),
    includes = [
        "include",
    ],
)
```
Bazel generates an XCode project with one target for each named rule.

The `grpc` target is of type `framework`, and declares a dependency on the `MyFramework_library`
target, of type `library.static`. The `MyFramework_binary` XCode target seems to not matter; it is
disconnected from the other two.

Bazel adds these build settings to each target:

```go
HEADER_SEARCH_PATHS = (
  "$(WORKSPACE_ROOT)/include",
  "$(WORKSPACE_ROOT)/bazel-out/local-ios_extension-fastbuild/genfiles/include",
  "$(inherited)",
);
USER_HEADER_SEARCH_PATHS = "$(WORKSPACE_ROOT) $(WORKSPACE_ROOT)/bazel-out/local-ios_extension-fastbuild/genfiles $(WORKSPACE_ROOT)";
WORKSPACE_ROOT = "/Users/jcanizales/git/grpc/src/objective-c/Bazel-Frameworks";
```

The `grpc` XCode target has three build phases: Frameworks, Sources, and Resources. The Resources
phase is empty and the Sources phase only has a dummy source file. Curiously, the Frameworks phase
includes `libMyFramework_library_.a`, which is the declared product of the `MyFramework_library`
target.

The `MyFramework_library` target has the same three build phases, but now only Sources is non-empty,
listing `mynestedsource.c`.

This setup allows the compiler to find both includes of `mynestedsource.c`:

- `#include <grpc/myheader.h>`, from the `HEADER_SEARCH_PATHS` of the static library target.
- `#include "nested/dir/mynestedheader.h"`, from its `USER_HEADER_SEARCH_PATHS`.

**Comment by jcanizales** [comment]
(This study is valid as of version 1.0.0 of Cocoapods).

From a Podspec like this, which is what we use for gRPC:

```ruby
s.name = 'MyFramework'
s.module_name = 'grpc'

s.source_files = 'include/grpc/myheader.h',
                 'nested/dir/mynestedheader.h',
                 'nested/dir/mynestedsource.c'
s.private_header_files = 'nested/dir/mynestedheader.h'

s.header_mappings_dir = '.'
# This isn't officially supported in Cocoapods. We've asked for an alternative:
# https://github.com/CocoaPods/CocoaPods/issues/4386
s.xcconfig = {
  'USE_HEADERMAP' => 'NO',
  'ALWAYS_SEARCH_USER_PATHS' => 'NO',
  'USER_HEADER_SEARCH_PATHS' => '"$(PODS_ROOT)/Headers/Private/MyFramework"',
  'HEADER_SEARCH_PATHS' => '"$(PODS_ROOT)/Headers/Private/MyFramework/include"'
}
```

Cocoapods generates an XCode project with a framework target named `MyFramework`, which generates
`grpc.framework` as a product. The target has five build phases:

- Sources,
- Frameworks,
- Headers,
- Copy include/grpc Public Headers, and
- Copy nested/dir Private Headers.

The Sources phase lists `mynestedsource.c`, the Headers phase lists both headers (probably
unnecessarily), and the Frameworks phase is trivial. The other two are "Copy Files" phases that copy
their headers into `$(PUBLIC_HEADERS_FOLDER_PATH)/include/grpc` and `$(PRIVATE_HEADERS_FOLDER_PATH)/nested/dir`
respectively. The macros turn out to be the `Headers` and `PrivateHeaders` directories inside the
built framework.

The target has the following build settings (in `Pods/Target Support Files/MyFramework/MyFramework.xcconfig`):

```go
HEADER_SEARCH_PATHS = "${PODS_ROOT}/Headers/Private" "${PODS_ROOT}/Headers/Public" "$(PODS_ROOT)/Headers/Private/MyFramework/include"
PODS_ROOT = ${SRCROOT}
USER_HEADER_SEARCH_PATHS = "$(PODS_ROOT)/Headers/Private/MyFramework"
```

Because the links under `${PODS_ROOT}/Headers` aren't generated when using frameworks, the compiler
can't find the includes of `mynestedsource.c`. To obtain an effect similar to Bazel's we would need
a variable with the path to the root of the pod sources (like Bazel's `WORKSPACE_ROOT`). We can't
assume it's going to be `$(PODS_ROOT)/MyFramework`, as that's only the case for pods downloaded from
the repo, not for local pods (what Cocoapods calls "development pods").

**Comment by jcanizales** [comment]
There are multiple ways of fixing Cocoapods to let it handle libraries with a directory structure like gRPC's.

One way would be to let the podspec developer specify what are the include paths, like Bazel or GYP do (https://github.com/CocoaPods/CocoaPods/issues/4386).

Another would be to add a build setting to the XCode target to point to the root of the pod sources, as suggested [here](https://github.com/CocoaPods/CocoaPods/issues/4386#issuecomment-147901203) (except that it'd need to vary for local development pods installed using `:path`). This should be a small change in this file https://github.com/CocoaPods/CocoaPods/blob/master/lib/cocoapods/generator/xcconfig/pod_xcconfig.rb#L42, and I might end up preparing a pull request for it.

The immediate solution I'm going to do, though, is to change the `USER_HEADER_SEARCH_PATHS` and `HEADER_SEARCH_PATHS` configured in gRPC's podspec to point to `$(PODS_ROOT)/podname` instead of `$(PODS_ROOT)/Headers`. This should solve the problem for the most common case, which is downloading the pod from Cocoapods' repo.

For the other use case, local development of gRPC, I'm going to rewrite those build settings on `pod install` by adding a post-install hook to our tests Podfile. [This is a nice example](https://github.com/CocoaPods/cocoapods-integration-specs/blob/master/install_podfile_callbacks/before/Podfile) of how to do it. It's not ideal, because it relies on where your Podfile is with respect to the cloned gRPC repo. But it unblocks everybody.

**Comment by jcanizales** [comment]
@makdharma FYI

**Comment by avf** [comment]
Hi @jcanizales, I'm trying to get gRPC to compile while using the `use_frameworks` flag and I've managed to get it to build by manually copying the generated `Headers` folder from `pod install` without `use_frameworks` and adding the paths to the `HEADER_SEARCH_PATHS` manually, like you suggest [here](https://github.com/CocoaPods/CocoaPods/issues/4605#issue-119331957). 

Unfortunately, while gRPC builds like this, I run into more problems later, since I'm trying to use gRPC as a subdependency in my own Swift framework, and as soon as I add any of the gRPC headers to my umbrella header, I get the `Include of non-modular header inside framework module` error. I think this was the reason why the CocoaPods devs decided to not generate the `Headers` directory (when `use_frameworks` is set) in the first place. Please see [this comment](https://github.com/CocoaPods/CocoaPods/issues/4420#issuecomment-151674437) and the corresponding issue for more information, the important bit is this:
> The compiler will run into issues if the same header file is accessible both through Header Search Paths (-I, -isystem) and Framework Search Paths (-F, -iframework), even if there are symbolic links involved. In these cases, you should prefer using Framework Search Paths.

Unfortunately, I'm not sure how to fix this problem - do you have any suggestion, even if it's a hacky fix/workaround? I've been trying to get this to work for a couple days now and I'm running out of ideas.

**Comment by jcanizales** [comment]
This is solved by #6849 and will be pushed to Cocoapods as soon as I can placate the @#$%#$^@ Cocoapods linter.

**Comment by pkieltyka** [comment]
@jcanizales thank you for all your efforts, that is awesome. I really believe the design of CocoaPods is flawed and its clearly very buggy. I think https://github.com/Carthage/Carthage is a better approach.

**Comment by fabware** [comment]
@jcanizales I've tested against V1.0.0 , it works, thanks for your effort.
